Sources/SourceKitLSP/Swift/SwiftLanguageServer.swift

```diff
@@ -161,6 +161,111 @@ public final class SwiftLanguageServer: ToolchainLanguageServer {
     }
   }
 
+  /// Updates the lexical and syntactic tokens for the given `snapshot`.
+  /// Must be called on `self.queue`.
+  private func updateLexicalAndSyntacticTokens(
+    response: SKDResponseDictionary,
+    for snapshot: DocumentSnapshot
+  ) {
+    dispatchPrecondition(condition: .onQueue(queue))
```

Comment:
Apologies if you've answered this already, I haven't read all the earlier comments on this PR, but have you explored the performance of this yet in a large file? It would be good to understand the impact on

* Performance of opening a large file - ideally > 5000 lines.
* Performance of making an edit in a large file (presumably fast with deltas)
* Performance of making an edit in a large file when using "full document" edits. Some editors may not have incremental edit support.
* Performance of updating the semantic tokens in a large file.

Reply:
I have experimented a bit with using the highlighting in big files (~ 6700 lines) with VSCode and it seems to be pretty responsive. Highlighting shows up in <1s and even when performing multiple edits in rapid succession, we get new tokens pretty quickly.

Performance for making edits in a large file with an editor that only supports full document edits would be similar to opening a new document each time, I think, i.e. slightly less responsive highlighting, but still with sub-second update times.

Do note, however, that while edits to the document are sent in deltas to the language server and SourceKit updates the lexical tokens (from syntaxmaps) in deltas, the updated tokens are transferred in full every time to the client (Although LSP's encoding seems pretty efficient, we might want to support delta responses in the future. This would require more sophisticated diffing logic on the server though, that I would consider out-of-scope for this PR).

Reply:
> Highlighting shows up in <1s and even when performing multiple edits in rapid succession, we get new tokens pretty quickly.

For opening the file this seems pretty good, but for edits I think we need more precise numbers.  I'm less concerned about the latency of getting the highlighting itself than the time spent processing and serializing the tokens where it blocks the primary server queues. Sub-second could be completely fine, or it could still be too slow if it is delaying code-completion.

To give some guidelines, I think anything over ~10 ms is bad and over ~100 ms is probably unacceptable if it's blocking the server queues after every edit, and especially if it "stacks up" work that is done redundantly when there are several edits in a row.  For example, when we are filtering down an existing code-completion as you type an identifier, that is something we are aiming for more like 10-20 ms total so that it could conceivably be 1-2 frames in the UI.

Drilling down a bit:
* Updating the tokens immediately after an edit. This seems performance critical since it likely blocks any possible code-completion requests. I think this needs to be <= 10 ms, or else we will need to move it to a background queue.

* Updating the tokens after the semantic update is received. This is less critical than the syntactic case, because edits get coalesced inside sourcekitd, and building the AST takes time, so it may _usually_ happen after any code-completion request. The processing time could still cause "jittery" behaviour where requests sometimes get delayed due to a semantic update from an earlier edit.  My guideline would be this should be < 100 ms if it's blocking the server queue, but it's possible we could get away with a bit more depending on the timing of the update and the other requests that are happening.

* Converting the tokens to their LSP format. The highlighting request itself is at least under the control of the editor, so in principle they can always choose to make the request for tokens when it won't block code-completion. I don't know if that's true in practice though, so we still should be mindful of the performance.

If it's a problem, we always have the option of moving the processing and conversion to lsp format of the tokens onto a background queue that doesn't block other requests, although it might make synchronization of state more complex.  So this all seems very solvable to me.

Reply:
I have profiled performing a bunch of edits in quick succession on a large (~6000 line) document and the main overhead involved seems to be parsing the tokens from SourceKit, especially due to expensive UTF-8 to UTF-16 conversions (which are, however, unavoidable):

[flamegraph.zip](https://github.com/apple/sourcekit-lsp/files/6885122/flamegraph.zip)

Logging the execution times of the methods that parse and convert the tokens while editing the large file confirms that these are potentially expensive operations if run synchronously:

<details>

```
2021-07-27 13:45:28.431 sourcekit-lsp[108644:b67fc700] updatedLexicalAndSyntacticTokens(response:for:) took 110.22 ms to execute
2021-07-27 13:45:28.738 sourcekit-lsp[108644:b67fc700] updatedLexicalAndSyntacticTokens(response:for:) took 113.87 ms to execute
2021-07-27 13:45:28.931 sourcekit-lsp[108644:b67fc700] updatedLexicalAndSyntacticTokens(response:for:) took 80.92 ms to execute
2021-07-27 13:45:29.133 sourcekit-lsp[108644:b67fc700] updatedLexicalAndSyntacticTokens(response:for:) took 85.72 ms to execute
2021-07-27 13:45:29.262 sourcekit-lsp[108644:b67fc700] updatedLexicalAndSyntacticTokens(response:for:) took 0.01 ms to execute
2021-07-27 13:45:29.521 sourcekit-lsp[108644:b67fc700] updatedLexicalAndSyntacticTokens(response:for:) took 0.01 ms to execute
2021-07-27 13:45:29.563 sourcekit-lsp[108644:b67fc700] updatedLexicalAndSyntacticTokens(response:for:) took 0.01 ms to execute
2021-07-27 13:45:29.673 sourcekit-lsp[108644:b67fc700] updatedLexicalAndSyntacticTokens(response:for:) took 83.71 ms to execute
2021-07-27 13:45:29.921 sourcekit-lsp[108644:b67fc700] updatedLexicalAndSyntacticTokens(response:for:) took 101.14 ms to execute
2021-07-27 13:45:30.103 sourcekit-lsp[108644:b67fc700] updatedLexicalAndSyntacticTokens(response:for:) took 103.39 ms to execute
2021-07-27 13:45:30.264 sourcekit-lsp[108644:b67fc700] updatedLexicalAndSyntacticTokens(response:for:) took 79.71 ms to execute
2021-07-27 13:45:30.774 sourcekit-lsp[108644:b67fc700] updatedSemanticTokens(response:for:) took 8.83 ms to execute
2021-07-27 13:45:32.685 sourcekit-lsp[108644:b67fc700] updatedLexicalAndSyntacticTokens(response:for:) took 0.01 ms to execute
2021-07-27 13:45:32.951 sourcekit-lsp[108644:b67fc700] updatedLexicalAndSyntacticTokens(response:for:) took 84.30 ms to execute
2021-07-27 13:45:33.123 sourcekit-lsp[108644:b67fc700] updatedLexicalAndSyntacticTokens(response:for:) took 88.88 ms to execute
2021-07-27 13:45:33.305 sourcekit-lsp[108644:b67fc700] updatedLexicalAndSyntacticTokens(response:for:) took 84.37 ms to execute
2021-07-27 13:45:33.491 sourcekit-lsp[108644:b67fc700] updatedLexicalAndSyntacticTokens(response:for:) took 0.01 ms to execute
```

</details>

So ideally we would want to do this asynchronously. I am, however, not sure how well that would play with the request flow for semantic tokens, we may have to await processing the (lexical/syntactic) tokens from the semanticTokens request handler, since we might otherwise end up with an outdated view of the document's tokens after the edit (and a potential race condition).

Reply:
Remind me: what exactly do we need the document sub-structure for in the token handling that isn't already part of the lexical tokens or the semantic tokens?  I'm assuming that's the expensive part of `updatedLexicalAndSyntacticTokens` since the lexical tokens are a delta update only.

> we may have to await processing the (lexical/syntactic) tokens from the semanticTokens request handler, since we might otherwise end up with an outdated view of the document's tokens after the edit (and a potential race condition).

We need the tokens to be range-shifted and possibly discarded if they overlap the edit, but that doesn't require them to be up to date semantically, and we would send the notification when that happens anyway.  Can we handle the range-shifting and handle the lexical tokens (which are only an incremental update) synchronously with the edit, but handle the rest asynchronously?

Reply:
We currently use the document substructure to provide accurate syntactic tokens e.g. for identifiers of declared variables, since we ignore lexical tokens of kind `keys.substructure.syntaxtype_identifier` (mostly because they are generally less specific than VSCode's default highlighting, causing tokens to 'flicker' between the semantic color and the general `syntaxtype_identifier` color, see https://github.com/apple/sourcekit-lsp/pull/414#discussion_r666275815). 

> Can we handle the range-shifting and handle the lexical tokens (which are only an incremental update) synchronously with the edit, but handle the rest asynchronously?

I think so, I'm currently looking into that.

Reply:
Traversing and parsing the syntactic tokens from the document substructure seems to be the most expensive part of `updatedLexicalAndSyntacticTokens`, if we only process lexical tokens after edits (synchronously), our processing time drops to 0-3 ms for edits in > 6000 line files.

Reply:
> if we only process lexical tokens after edits (synchronously), our processing time drops to 0-3 ms for edits in > 6000 line files.

Nice!

Reply:
Another option would be to simply skip syntactic tokens if the file length exceeds a certain threshold. Since only a small subset of the tokens are syntactic (mostly just the identifiers of declared variables), I think it wouldn't look too different either. What do you think?

Reply:
I would prefer not to have that kind of limit.  If it was a really large limit that you wouldn't hit before maybe 20,000 lines I think that might be okay, but we're talking about fairly realistic file sizes.  In the best case it will result in inconsistent behaviour between small and large files.

> I think it wouldn't look too different either.

If users notice the difference it will feel buggy, and if they don't notice the difference then arguably we could just not include the syntactic tokens at all.  I would argue we either make it fast enough to always have enabled or we skip the syntactic tokens always.

Reply:
Note: it's always an option to change our mind later too, so we could disable these tokens for now and put in the work to enable them later if we think it's worth it.